src/lib/es5.d.ts

```diff
@@ -1273,7 +1273,20 @@ declare type PropertyDecorator = (target: Object, propertyKey: string | symbol)
 declare type MethodDecorator = <T>(target: Object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>) => TypedPropertyDescriptor<T> | void;
 declare type ParameterDecorator = (target: Object, propertyKey: string | symbol, parameterIndex: number) => void;
 
-declare type PromiseConstructorLike = new <T>(executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void) => PromiseLike<T>;
+/** An object type that can be definitely be awaited. Do not inherit from this type. */
+declare type Awaitable<T> = { then(onfulfilled: (value: T) => any): any; };
+
+/** An object type that may not be awaitable. Do not inherit from this type. */
+declare type NonAwaitable = { then(...args: any[]): any; };
+
+/** Gets the type resulting from awaiting `T`. This does **not** recursively unwrap nested promises. */
+declare type Awaited<T> =
+    T extends Awaitable<Awaitable<infer U>> ? U :

```

Comment:
Do we figure that two levels deep is enough for most uses, and if anyone complains, we can always add more later?

Reply:
Duh, two levels of conditional type inference should be enough for anyone™

Reply:
I guess if we currently don't do better than two levels, this will do for now.

Reply:
If you are using the `Promise` definition from our libs then you won't need anything deeper. Once this is shipping we can modify the definitions in DT to use it as well. Per my examples above: `type T3 = Awaited<Promise<Promise<number>>>; // number` for any level of nesting using either `Promise` or `PromiseLike`.

Reply:
We actually only need:
```ts
declare type Awaited<T> =
    T extends Awaitable<infer U> ? U :
    T extends { then(...args: any[]): any; } ? never :
    T;
```

I seemed to have forgotten to stage an additional change to es5.d.ts.